metals/src/main/scala/scala/meta/internal/metals/codeactions/ExtractRenameMember.scala

```diff
@@ -0,0 +1,338 @@
+      } else Nil
+    }
+  }
+
+  private def extractClassAction(
+      uri: String,
+      member: Member
+  ): l.CodeAction = {
+
+    val memberName = member.name.value
+    val range = member.name.pos.toLSP
+
+    val codeAction = new l.CodeAction()
+    codeAction.setTitle(
+      ExtractRenameMember.title(getMemberType(member), memberName)
+    )
+    codeAction.setKind(kind)
+    codeAction.setCommand(
+      ServerCommands.ExtractMemberDefinition.toLSP(
+        List(
+          uri,
+          range.getStart.getLine(): java.lang.Integer,
+          range.getStart.getCharacter(): java.lang.Integer
+        )
+      )
+    )
+
+    codeAction
+  }
+
+  private def newURIFromClass(uri: String, member: Member): String = {
+    val srcFileName = uri.toAbsolutePath.filename
+    val classDefnName = member.name.value
+    uri.replace(srcFileName, s"$classDefnName.scala")
+  }
+
+  override def kind: String = l.CodeActionKind.RefactorExtract
+
+  override def executeCommand(token: CancelToken)(implicit
+      ec: ExecutionContext
+  ): PartialFunction[CodeActionCommandData, Future[
+    Option[CodeActionCommandResult]
+  ]] = {
+    case ExtractMemberDefinitionData(
+          uri,
+          params,
+          ExtractRenameMember.extractDefCommandDataType
+        ) => {
+      def isCompanion(
+          member: Member
+      )(candidateCompanion: Member): Boolean = {
+        val differentMemberWithSameName = !candidateCompanion.equals(member) &&
+          candidateCompanion.name.value.equals(member.name.value)
+        member match {
+          case _: Defn.Object => differentMemberWithSameName
+          case _ =>
+            candidateCompanion
+              .isInstanceOf[Defn.Object] && differentMemberWithSameName
+        }
+      }
+
+      val pos = params.getPosition
+      val range = new l.Range(pos, pos)
+      val path = uri.toAbsolutePath
+
+      val futureOpt = for {
+        tree <- trees.get(path)
+        definitions = membersDefinitions(tree)
+        memberDefn <- definitions.find(_.name.pos.toLSP.overlapsWith(range))
+        companion = definitions.find(isCompanion(memberDefn))
+        newFileContent = getNewFileContent(
+          tree,
+          range,
+          memberDefn,
+          companion
+        )
+        newFileUri = newURIFromClass(uri, memberDefn)
+        newFileMemberRange = new l.Range()
+      } yield for {
+        edits <- extractClassCommand(
+          newFileUri,
+          newFileContent,
+          memberDefn,
+          companion
+        )
+      } yield Option(
+        ExtractMemberDefinitionResult(edits, newFileUri, newFileMemberRange)
+      )
+
+      futureOpt.getOrElse(Future { None })
+    }
+  }
+}
+
+object ExtractRenameMember {
+  def getMemberType(member: Member): String = member match {
+    case _: Defn.Class => "class"
+    case _: Defn.Enum => "enum"
+    case _: Defn.Trait => "trait"
+    case _: Defn.Object => "object"
+    case _ => "unknown type"
```

Comment:
This function is used essentially only to get the right type in the CodeAction Title, and the Code Action itself cannot be created if the `Member` is not one from `class`/`enum`/`trait`/`object`, this "match map" is there just to avoid having "Extract member X from file Y.scala`, so even using an option at a certain point we would need to get it out of it and return some other string to "fill the space", or not showing anything in the CodeAction Title.

Reply:
I would just not show the code action at all if cannot create a title. In that case it means something went really wrong.